Event driven refresh

mareklibra opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ovirt/ovirt-web-ui/390)
<!-- Reviewable:end -->

jniederm commented:
Review status: 0 of 20 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/ovirtapi.js, line 301 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyMqnSlzG5BxQoOqdCx:-KyMqnSlzG5BxQoOqdCy:b-deeggp) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/ovirtapi.js#L301)):*
> ```JavaScript
>       cpu: {
>         topology: {
>           cores: template.cpu && template.cpu.topology ? template.cpu.topology.cores : undefined,
> ```

I guess `template.cpu && template.cpu.topology && template.cpu.topology.cores` does the same.

---

*[src/ovirtapi.js, line 410 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyMrf9IvVprblJQ5G6u:-KyMrf9IvVprblJQ5G6v:b-nt6zpe) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/ovirtapi.js#L410)):*
> ```JavaScript
> 
>   // ----
>   getEvents ({ lastEventIndexReceived }: { lastEventIndexReceived: ?number }): Promise<Object> {
> ```

Or maybe `lastReceivedEventIndex`?

---

*[src/sagas.js, line 499 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyN3adwALYFTmI-Npdx:-KyN3adxqthcgfJm-SPL:b-tt9a12) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/sagas.js#L499)):*
> ```JavaScript
> }
> 
> function* fetchSingleResource (action, apiMethod, toInternalConvertor, addAction, removeAction) {
> ```

Maybe rather `toInternalConverter`?

---

*[src/sagas.js, line 500 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyN7jK3ucJHFI-mMK4c:-KyN7jK3ucJHFI-mMK4d:bulv0xn) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/sagas.js#L500)):*
> ```JavaScript
> 
> function* fetchSingleResource (action, apiMethod, toInternalConvertor, addAction, removeAction) {
>   const res = yield callExternalAction(apiMethod, Api[apiMethod], action)
> ```

Type-safety and IDE clickability can be improved by passing actual function in `apiMethod` param instead of function name:

```
const res = yield callExternalAction(apiMethod.name, apiMethod.bind(Api), action)
```

---

*[src/actions/events.js, line 13 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyN98u1OgSPoKMhL2NW:-KyN98u1OgSPoKMhL2NX:bb7od0r) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/actions/events.js#L13)):*
> ```JavaScript
> }
> 
> export function getEvents ({ lastEventIndexReceived }) {
> ```

same

---

*[src/saga/utils.js, line 96 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyNGC1UeYmkzUepWWjD:-KyNGC1UeYmkzUepWWjE:b-h1e1id) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/saga/utils.js#L96)):*
> ```JavaScript
> 
> export function * foreach (array, fn, context) {
>   if (!array) {
> ```

I'm not a fan of such solution - silently do nothing in case of wrong argument. It goes against "fail fast" principle. Can we just not call the generator method if `array` is falsy? 

---

*[src/saga/utils.js, line 121 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyNDOhb5rAksY7-BebI:-KyNDOhcMdVBFWeFaQth:b6l1yg9) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/saga/utils.js#L121)):*
> ```JavaScript
> export function compareVersion (actual, required) {
>   logDebug(`compareVersion(), actual=${JSON.stringify(actual)}, required=${JSON.stringify(required)}`)
> ```

I guess

```
if (actual.major === required.major && actual.minor < required.minor) {
  return false
}
return actual.major >= required.major
```

does the same.

---

*[src/saga/utils.js, line 137 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyNEM41IBHE1ZnHJTxi:-KyNEM42E6Y5RH5Gxo2h:br0h6ue) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/saga/utils.js#L137)):*
> ```JavaScript
>  * Wait for predicate() to become true.
>  */
> export function * waitForIt (predicate) {
> ```

How about just `waitFor` so it can be read "wait for predicate"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390)*
<!-- Sent from Reviewable.io -->

jniederm commented:
Generally it makes me sad to see combining of naming pattern get**Single**Entity and removeEntity.  I found example of both existing before this patch. Nevertheless I'd suggest to select one pattern and try to follow it in whole project.

mareklibra commented:
Review status: 0 of 20 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/ovirtapi.js, line 301 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyMqnSlzG5BxQoOqdCx:-KyQ2n0Rycu0qjb0UThu:bpx9nm6) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/ovirtapi.js#L301)):*
<details><summary><i>Previously, jniederm wrote…</i></summary><blockquote>

I guess `template.cpu && template.cpu.topology && template.cpu.topology.cores` does the same.
</blockquote></details>

This might be `false` but there's no semantic difference in our case.

---

*[src/ovirtapi.js, line 410 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyMrf9IvVprblJQ5G6u:-KyQ3qMFceAFKs_muxzJ:b32wtd) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/ovirtapi.js#L410)):*
<details><summary><i>Previously, jniederm wrote…</i></summary><blockquote>

Or maybe `lastReceivedEventIndex`?
</blockquote></details>

ok

---

*[src/sagas.js, line 499 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyN3adwALYFTmI-Npdx:-KyQ3uRNZFLpjiIgvVdf:b4vxylh) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/sagas.js#L499)):*
> toInternalConverter
ok

---

*[src/saga/utils.js, line 96 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyNGC1UeYmkzUepWWjD:-KyQ5oUv0VDGl2n1al5V:b-cns713) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/saga/utils.js#L96)):*
<details><summary><i>Previously, jniederm wrote…</i></summary><blockquote>

I'm not a fan of such solution - silently do nothing in case of wrong argument. It goes against "fail fast" principle. Can we just not call the generator method if `array` is falsy? 
</blockquote></details>

right, fixed

---

*[src/saga/utils.js, line 121 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyNDOhb5rAksY7-BebI:-KyQ6Y1T9RZng4bQrnkp:b-uzybr5) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/saga/utils.js#L121)):*
> if (actual.major === required.major && actual.minor < required.minor) {
>   return false
> }
> return actual.major >= required.major
right, simplified

---

*[src/saga/utils.js, line 137 at r1](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyNEM41IBHE1ZnHJTxi:-KyQ6hL0IJAUldO7n-My:bnduuqn) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/40fee712440466de0b2ad242d7a6184c733aff99/src/saga/utils.js#L137)):*
>  can be read "w
I like waitForIt more :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390)*
<!-- Sent from Reviewable.io -->

mareklibra commented:
Review status: 0 of 20 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[src/saga/events.js, line 86 at r2](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyKgIZ2e34gq3EucteI-r2-86:-KyVayNEOKIXbo2AguEm:bcbz1ac) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/ad347c66e4ba873c0cbd11735b9e59d97263f3d9/src/saga/events.js#L86)):*
<details><summary><i>Previously, jniederm wrote…</i></summary><blockquote>

Is there any reason why this is a generator function instead of regular one?
</blockquote></details>

Not any more, so simplifying

---

*[src/saga/events.js, line 111 at r2](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390#-KyKgIZ2e34gq3EucteI-r2-111:-KyV_DPOA_hhepmSjWgE:bjn5f57) ([raw file](https://github.com/ovirt/ovirt-web-ui/blob/ad347c66e4ba873c0cbd11735b9e59d97263f3d9/src/saga/events.js#L111)):*
<details><summary><i>Previously, jniederm wrote…</i></summary><blockquote>

I guess `Object.keys` would be more suitable here - it returns enumerable properties only.

```
Object.keys('a')                // [ '0' ] 
Object.getOwnPropertyNames('a') // [ '0', 'length' ]
```

Similar for other lines.
</blockquote></details>

right, thanks, too hasty

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ovirt/ovirt-web-ui/390)*
<!-- Sent from Reviewable.io -->

mareklibra commented:
>  getSingleEntity, addEntity, removeEntity naming convention
The *Single* improves readability and helps to avoid mistakes.
Anyway, I can imagine extension of these methods to accept an array entities, so let's remove the *Single* from names.

>  passing function reference instead of function name in src/sagas.js
done

jniederm commented on src/sagas.js:
```diff
@@ -487,11 +492,29 @@ function* fetchVmSessions ({ vmId }) {
+  yield fetchPool(getPool({ poolId: action.payload.poolId }))
+}
+
+function* fetchResource (action, methodName, method, toInternalConverter, addAction, removeAction) {
```
> >  passing function reference instead of function name in src/sagas.js

> done

It doesn't look like done to me.

mareklibra commented:
Merge is pending till 1.4 to mitigate risks for oVirt 4.2.

mareklibra commented:
If possible, let's avoid branching of the project

jelkosz commented:
please note there might be issues using this with user which does not have any admin role due to https://bugzilla.redhat.com/show_bug.cgi?id=1469055

mareklibra commented:
oVirt 4.2 is out so unblocking.

Will require extensive rebase and testing, former context must not be valid now.

gregsheremeta commented:
closing for now, may revisit
